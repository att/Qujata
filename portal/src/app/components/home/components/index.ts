export * from './experiment';
export * from './latest-experiments';
