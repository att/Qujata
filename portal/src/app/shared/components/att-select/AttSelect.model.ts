// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface AttSelectOption<T = any> {
    value: string;
    label: string;
    metadata?: T;
    isDisabled?: boolean,
}

export enum AttSelectTheme {
    REGULAR = 'regular',
    PRIMARY = 'primary',
    WRAPPER = 'wrapper',
}
