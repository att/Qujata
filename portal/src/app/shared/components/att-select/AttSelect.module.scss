@import 'src/styles/variables-keys';
@import './AttSelect.reference';

$attSelectSpinnerSizeValue: var($selectSpinnerSize, 18px);
$attSelectSpinnerRightMarginValue: var($selectSpinnerRightMargin, 10px);
$attSelectSpinnerLeftMarginValue: var($selectSpinnerLeftMargin, 2px);

.att_select_primary_theme {
  #{$selectBorderColor}: var($attCobalt);
  #{$selectControlBackgroundColor}: transparent;
  #{$selectPlaceholderColor}: var($attCobalt);
  #{$selectIndicatorsColor}: var($attCobalt);
  &:hover {
    #{$selectControlBackgroundColor}: var($attCobalt);
    #{$selectPlaceholderColor}: var($primaryWhite);
    #{$selectIndicatorsColor}: var($primaryWhite);
  }
}

.att_select_wrapper_theme {
  #{$selectControlBackgroundColor}: transparent;
  #{$selectBorderColor}: transparent;
}

.att_select_error {
  #{$selectBorderColor}: var($errorTextColor);
}

.att_select_spinner {
  block-size: $attSelectSpinnerSizeValue;
  inline-size: $attSelectSpinnerSizeValue;
  margin-inline-end: $attSelectSpinnerRightMarginValue;
  margin-inline-start: $attSelectSpinnerLeftMarginValue;
}
