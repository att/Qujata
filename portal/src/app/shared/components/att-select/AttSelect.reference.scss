$selectSpinnerSize: --att-select-spinner-size;
$selectSpinnerRightMargin: --att-select-spinner-right-margin;
$selectSpinnerLeftMargin: --att-select-spinner-left-margin;
