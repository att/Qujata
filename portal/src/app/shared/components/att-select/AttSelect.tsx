/* eslint-disable react/no-unused-prop-types */
/* eslint-disable @typescript-eslint/ban-types */
import cn from 'classnames';
import React, { FocusEventHandler, ForwardedRef, forwardRef, ForwardRefExoticComponent, PropsWithChildren, PropsWithoutRef, RefAttributes } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import Select, { ActionMeta, Options, SelectInstance } from 'react-select';
import styles from './AttSelect.module.scss';
import { AttSelectOption, AttSelectTheme } from './AttSelect.model';
import { Spinner, SpinnerSize } from '../att-spinner';
import { AttSelectCssClassByTheme } from './AttSelect.const';

export type OnSelectChangedCallbackType = (newValue: unknown, actionMeta: ActionMeta<unknown>) => void;
export type OnSelectChangedType = AttSelectOption | Options<AttSelectOption> | null;
export type OnSelectChanged = (event: OnSelectChangedType) => void;

export interface AttSelectProps {
  options: AttSelectOption[];
  value: AttSelectOption | AttSelectOption[] | undefined | null;
  required?: boolean;
  defaultValue?: AttSelectOption | AttSelectOption[];
  onChange: OnSelectChanged;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  isProcessing?: boolean;
  disabled?: boolean;
  readonly?: boolean;
  isClearable?: boolean;
  hasError?: boolean;
  className?: string;
  placeholder?: string;
  customComponent?: {
    Option?: React.FC;
    ValueContainer?: React.FC;
    DropdownIndicator?: React.FC;
    MenuList?: React.FC;
  };
  theme?: AttSelectTheme;
  isMulti?: boolean;
  closeMenuOnSelect?: boolean;
  hideSelectedOptions?: boolean;
  e2eId?: string;
  id?: string;
}

export type AttSelectExposeRef = SelectInstance;
type AttSelectPrivateType = React.ForwardRefRenderFunction<AttSelectExposeRef, AttSelectProps>;
const AttSelectPrivate: AttSelectPrivateType = (props: AttSelectProps, ref: ForwardedRef<AttSelectExposeRef>) => (
  <Select
    ref={ref}
    className={getAttSelectClassName(props)}
    classNamePrefix='att_select'
    options={props.options}
    onChange={props.onChange as OnSelectChangedCallbackType}
    onBlur={props.onBlur}
    isDisabled={props.disabled || props.readonly}
    isLoading={props.isProcessing}
    defaultValue={props.defaultValue}
    value={props.value}
    placeholder={props.placeholder}
    isMulti={props.isMulti}
    isClearable={props.isClearable}
    // menuIsOpen={true}
    closeMenuOnSelect={props.closeMenuOnSelect}
    hideSelectedOptions={props.hideSelectedOptions}
    required={props.required}
    components={{
      IndicatorSeparator,
      LoadingIndicator,
      ...(props.customComponent || {}),
    }}
    id={props.e2eId}
    inputId={props.id}
  />
);

function getAttSelectClassName(props: AttSelectProps): string {
  const { className, readonly, isProcessing, hasError, isMulti } = props;
  const theme: AttSelectTheme = props.theme || AttSelectTheme.REGULAR;
  return cn(
    'att_select_container',
    className,
    AttSelectCssClassByTheme.get(theme),
    {
      'att_select_container--is-readonly': readonly,
      'att_select_container--is-loading': isProcessing,
      att_select_multi_container: isMulti,
      [styles.att_select_error]: hasError,
    },
  );
}
function IndicatorSeparator() {
  // eslint-disable-next-line no-null/no-null
  return null;
}

function LoadingIndicator() {
  return <Spinner size={SpinnerSize.NONE} className={styles.att_select_spinner} />;
}

type AttSelectRef = ForwardRefExoticComponent<PropsWithoutRef<PropsWithChildren<AttSelectProps>> & RefAttributes<AttSelectExposeRef>>
export const AttSelect: AttSelectRef = forwardRef<AttSelectExposeRef, PropsWithChildren<AttSelectProps>>(AttSelectPrivate);
