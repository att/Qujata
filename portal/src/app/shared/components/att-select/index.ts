export * from './AttSelect';
export * from './AttSelect.model';
