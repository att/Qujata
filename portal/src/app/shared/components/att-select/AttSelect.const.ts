import styles from './AttSelect.module.scss';
import { AttSelectTheme } from './AttSelect.model';

export const AttSelectCssClassByTheme: Map<AttSelectTheme, string> = new Map<AttSelectTheme, string>([
  [AttSelectTheme.PRIMARY, styles.att_select_primary_theme],
  [AttSelectTheme.WRAPPER, styles.att_select_wrapper_theme],
  [AttSelectTheme.REGULAR, ''],
]);
