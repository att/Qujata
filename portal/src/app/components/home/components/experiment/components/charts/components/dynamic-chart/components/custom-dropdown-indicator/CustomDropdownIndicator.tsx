import { useCallback } from "react";
import { DropdownIndicatorProps, components } from "react-select";
import { AttSelectOption } from "../../../../../../../../../../shared/components/att-select";
import { ReactComponent as ArrowDownSelectorSvg } from '../../../../../../../../../../../assets/images/arrow-down-selector.svg';

export const CustomDropdownIndicator: React.FC<DropdownIndicatorProps<AttSelectOption, any>> = (props) => {
    const { selectProps } = props;

    const handleClick: () => void = useCallback((): void => {
        if (selectProps.menuIsOpen) {
            selectProps.onMenuClose();
        } else {
            selectProps.onMenuOpen();
        }
    }, [selectProps]);
    
    return (
        <div onClick={handleClick} >
            <components.DropdownIndicator {...props}>
                <ArrowDownSelectorSvg />
            </components.DropdownIndicator>
        </div>
    );
};
