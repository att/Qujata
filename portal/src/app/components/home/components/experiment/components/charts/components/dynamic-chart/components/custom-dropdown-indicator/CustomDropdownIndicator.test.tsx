import { fireEvent, render, RenderResult } from '@testing-library/react';
import { CustomDropdownIndicator } from './CustomDropdownIndicator';
import { DropdownIndicatorProps } from 'react-select';
import { AttSelectOption } from '../../../../../../../../../../shared/components/att-select';

describe('CustomDropdownIndicator', () => {
  const mockProps: DropdownIndicatorProps<AttSelectOption, any> = {
      innerProps: undefined as any,
      isFocused: false,
      isDisabled: false,
      clearValue: jest.fn(),
      cx: jest.fn(),
      getStyles: jest.fn(),
      getClassNames: jest.fn(),
      getValue: jest.fn(),
      hasValue: false,
      isMulti: false,
      isRtl: false,
      options: [],
      selectOption: jest.fn(),
      selectProps: { menuIsOpen: false, onMenuOpen: jest.fn(), onMenuClose: jest.fn() } as any,
      setValue: jest.fn(),
      theme: undefined as any,
  };

  test('should render CustomDropdownIndicator when the menu is close', () => {
    const { container, getByTestId }: RenderResult = render(<CustomDropdownIndicator {...mockProps} />);

    const dropdownIndicatorWrapperElement: HTMLElement = getByTestId('dropdown_indicator_wrapper');
    fireEvent.click(dropdownIndicatorWrapperElement);

    expect(container).toBeTruthy();
    expect(mockProps.selectProps.onMenuOpen).toHaveBeenCalled();
    expect(mockProps.selectProps.onMenuClose).not.toHaveBeenCalled();
  });

  test('should render CustomDropdownIndicator when the menu is open', () => {
    mockProps.selectProps.menuIsOpen = true;
    const { container, getByTestId }: RenderResult = render(<CustomDropdownIndicator {...mockProps} />);

    const dropdownIndicatorWrapperElement: HTMLElement = getByTestId('dropdown_indicator_wrapper');
    fireEvent.click(dropdownIndicatorWrapperElement);

    expect(container).toBeTruthy();
    expect(mockProps.selectProps.onMenuOpen).not.toHaveBeenCalled();
    expect(mockProps.selectProps.onMenuClose).toHaveBeenCalled();
  });
});
