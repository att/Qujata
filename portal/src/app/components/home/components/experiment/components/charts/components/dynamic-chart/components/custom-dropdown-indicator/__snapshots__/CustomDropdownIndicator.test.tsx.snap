// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDropdownIndicator should render CustomDropdownIndicator 1`] = `
<div>
  <div
    class="css-0"
  >
    <svg>
      arrow-down-selector.svg
    </svg>
  </div>
</div>
`;
